// @microsoft/sp-tslint-rules/lib/utils/implicitRoles/index.js
"use strict";
var a_1 = require('./a');
exports.A = a_1.A;
var area_1 = require('./area');
exports.AREA = area_1.AREA;
var article_1 = require('./article');
exports.ARTICLE = article_1.ARTICLE;
var aside_1 = require('./aside');
exports.ASIDE = aside_1.ASIDE;
var body_1 = require('./body');
exports.BODY = body_1.BODY;
var button_1 = require('./button');
exports.BUTTON = button_1.BUTTON;
var datalist_1 = require('./datalist');
exports.DATALIST = datalist_1.DATALIST;
var details_1 = require('./details');
exports.DETAILS = details_1.DETAILS;
var dialog_1 = require('./dialog');
exports.DIALOG = dialog_1.DIALOG;
var dl_1 = require('./dl');
exports.DL = dl_1.DL;
var form_1 = require('./form');
exports.FORM = form_1.FORM;
var h1_1 = require('./h1');
exports.H1 = h1_1.H1;
var h2_1 = require('./h2');
exports.H2 = h2_1.H2;
var h3_1 = require('./h3');
exports.H3 = h3_1.H3;
var h4_1 = require('./h4');
exports.H4 = h4_1.H4;
var h5_1 = require('./h5');
exports.H5 = h5_1.H5;
var h6_1 = require('./h6');
exports.H6 = h6_1.H6;
var hr_1 = require('./hr');
exports.HR = hr_1.HR;
var img_1 = require('./img');
exports.IMG = img_1.IMG;
var input_1 = require('./input');
exports.INPUT = input_1.INPUT;
var li_1 = require('./li');
exports.LI = li_1.LI;
var link_1 = require('./link');
exports.LINK = link_1.LINK;
var menu_1 = require('./menu');
exports.MENU = menu_1.MENU;
var menuitem_1 = require('./menuitem');
exports.MENUITEM = menuitem_1.MENUITEM;
var meter_1 = require('./meter');
exports.METER = meter_1.METER;
var nav_1 = require('./nav');
exports.NAV = nav_1.NAV;
var ol_1 = require('./ol');
exports.OL = ol_1.OL;
var option_1 = require('./option');
exports.OPTION = option_1.OPTION;
var output_1 = require('./output');
exports.OUTPUT = output_1.OUTPUT;
var progress_1 = require('./progress');
exports.PROGRESS = progress_1.PROGRESS;
var section_1 = require('./section');
exports.SECTION = section_1.SECTION;
var select_1 = require('./select');
exports.SELECT = select_1.SELECT;
var tbody_1 = require('./tbody');
exports.TBODY = tbody_1.TBODY;
var textarea_1 = require('./textarea');
exports.TEXTAREA = textarea_1.TEXTAREA;
var tfoot_1 = require('./tfoot');
exports.TFOOT = tfoot_1.TFOOT;
var thead_1 = require('./thead');
exports.THEAD = thead_1.THEAD;
var ul_1 = require('./ul');
exports.UL = ul_1.UL;
