// @microsoft/sp-build-web/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
// Imports
var fs = require('fs');
var os = require('os');
var path = require('path');
var stripJsonComments = require('strip-json-comments');
var Validator = require('z-schema');
var yargs = require('yargs');
var sp_tslint_rules_1 = require('@microsoft/sp-tslint-rules');
var gulp_core_build_1 = require('@microsoft/gulp-core-build');
var gulpCoreBuild = require('@microsoft/gulp-core-build');
var gulp_core_build_typescript_1 = require('@microsoft/gulp-core-build-typescript');
var gulp_core_build_sass_1 = require('@microsoft/gulp-core-build-sass');
var gulp_core_build_serve_1 = require('@microsoft/gulp-core-build-serve');
var gulp_core_build_webpack_1 = require('@microsoft/gulp-core-build-webpack');
var gulp_core_build_karma_1 = require('@microsoft/gulp-core-build-karma');
var spBuildCoreTasks = require('@microsoft/sp-build-core-tasks');
// Exports
__export(require('@microsoft/gulp-core-build'));
__export(require('@microsoft/gulp-core-build-typescript'));
__export(require('@microsoft/gulp-core-build-sass'));
__export(require('@microsoft/gulp-core-build-serve'));
__export(require('@microsoft/gulp-core-build-webpack'));
__export(require('@microsoft/gulp-core-build-karma'));
__export(require('@microsoft/sp-build-core-tasks'));
exports.preCopy = new gulp_core_build_1.CopyTask();
exports.postCopy = new gulp_core_build_1.CopyTask();
exports.preCopy.name = 'pre-copy';
exports.postCopy.name = 'post-copy';
var enableCasperTests = true;
var schemaKey = '$schema';
var testTaskName = 'test';
var buildTaskName = 'build';
var bundleTaskName = 'bundle';
var deployAzureStorageTaskName = 'deploy-azure-storage';
var packageSolutionTaskName = 'package-solution';
var serveTaskName = 'serve';
var defaultTaskName = 'default';
var debugBuildMode = 'debug';
var shipBuildMode = 'ship';
function setBuildProperties(properties) {
    gulp_core_build_1.setConfig({ properties: properties });
}
exports.setBuildProperties = setBuildProperties;
var additionalBuildTasks = [];
/**
 * Add additional build tasks to run after the typescript tasks.
 */
function addBuildTasks(tasks) {
    if (tasks.length) {
        additionalBuildTasks.push.call(this, tasks);
    }
    else {
        additionalBuildTasks.push(tasks);
    }
}
exports.addBuildTasks = addBuildTasks;
function getSchemaFilePath(filename) {
    return path.join(__dirname, 'schemas', filename);
}
var SpWebBuild = (function () {
    function SpWebBuild() {
        this.customConfigs = [
            {
                callback: setBuildProperties,
                configFile: 'config.json',
                schemaFile: getSchemaFilePath('config.schema.json')
            },
            {
                callback: gulp_core_build_serve_1.serve.setConfig.bind(gulp_core_build_serve_1.serve),
                configFile: 'serve.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-serve#servetask',
                schemaFile: undefined
            },
            {
                callback: spBuildCoreTasks.deployAzureStorage.setConfig.bind(spBuildCoreTasks.deployAzureStorage),
                configFile: 'deploy-azure-storage.json',
                readmeUrl: 'https://github.com/OfficeDev/SharePointUXFramework/wiki/' +
                    'Notes-on-Azure-Deployment#configuration',
                schemaFile: undefined
            },
            {
                callback: spBuildCoreTasks.packageSolution.setConfig.bind(spBuildCoreTasks.packageSolution),
                configFile: 'package-solution.json',
                readmeUrl: 'https://github.com/OfficeDev/SharePointUXFramework/wiki/Notes-on-Solution-Packaging#packaging',
                schemaFile: undefined
            },
            {
                callback: exports.preCopy.setConfig.bind(exports.preCopy),
                configFile: 'pre-copy.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build/blob/master/src/CopyTask.ts',
                schemaFile: getSchemaFilePath('copy.schema.json')
            },
            {
                callback: exports.postCopy.setConfig.bind(exports.postCopy),
                configFile: 'post-copy.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build/blob/master/src/CopyTask.ts',
                schemaFile: getSchemaFilePath('copy.schema.json')
            },
            {
                callback: gulp_core_build_typescript_1.typescript.setConfig.bind(gulp_core_build_typescript_1.typescript),
                configFile: 'typescript.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-typescript/blob/master/src/TypeScriptTask.ts',
                schemaFile: undefined
            },
            {
                callback: gulp_core_build_typescript_1.tslint.setConfig.bind(gulp_core_build_typescript_1.tslint),
                configFile: 'tslint.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-typescript/blob/master/src/TSLintTask.ts',
                schemaFile: undefined
            },
            {
                callback: gulp_core_build_sass_1.default.setConfig.bind(gulp_core_build_sass_1.default),
                configFile: 'sass.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-sass/blob/master/src/SassTask.ts',
                schemaFile: undefined
            },
            {
                callback: gulp_core_build_serve_1.reload.setConfig.bind(gulp_core_build_serve_1.reload),
                configFile: 'reload.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-serve/blob/master/src/ReloadTask.ts',
                schemaFile: undefined
            },
            {
                callback: gulp_core_build_webpack_1.default.setConfig.bind(gulp_core_build_webpack_1.default),
                configFile: 'webpack.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-webpack/blob/master/src/WebpackTask.ts',
                schemaFile: undefined
            },
            {
                callback: gulp_core_build_karma_1.default.setConfig.bind(gulp_core_build_karma_1.default),
                configFile: 'karma.json',
                readmeUrl: 'https://github.com/Microsoft/gulp-core-build-karma/blob/master/src/KarmaTask.ts',
                schemaFile: undefined
            },
            {
                callback: spBuildCoreTasks.configureWebpack.setConfig.bind(spBuildCoreTasks.configureWebpack),
                configFile: 'configure-webpack.json',
                readmeUrl: 'https://www.npmjs.com/package/@microsoft/sp-build-core-tasks',
                schemaFile: undefined
            },
            {
                callback: spBuildCoreTasks.copyAssets.setConfig.bind(spBuildCoreTasks.copyAssets),
                configFile: 'prepare-deploy.json',
                readmeUrl: 'https://www.npmjs.com/package/@microsoft/sp-build-core-tasks',
                schemaFile: undefined
            },
            {
                callback: spBuildCoreTasks.writeManifests.setConfig.bind(spBuildCoreTasks.writeManifests),
                configFile: 'write-manifests.json',
                readmeUrl: 'https://www.npmjs.com/package/@microsoft/sp-build-core-tasks',
                schemaFile: undefined
            },
            {
                callback: spBuildCoreTasks.copyStaticAssets.setConfig.bind(spBuildCoreTasks.copyStaticAssets),
                configFile: 'copy-static-assets.json',
                readmeUrl: 'https://www.npmjs.com/package/@microsoft/sp-build-core-tasks',
                schemaFile: getSchemaFilePath('copy-static-assets.schema.json')
            }
        ];
        this._schemaValidator = new Validator({
            breakOnFirstError: true,
            noExtraKeywords: true,
            noTypeless: true
        });
    }
    SpWebBuild.prototype.initialize = function (gulp) {
        var yargsArgs = this.getYargs();
        var command = yargsArgs.argv._[0];
        exports.args = yargsArgs.argv;
        // TODO: VSO#178074: this code should be removed once we determine the interface for argument parsing with GCB
        switch (command) {
            case testTaskName:
                yargsArgs.reset()
                    .option('debug', {
                    describe: 'run tests in debug mode'
                })
                    .option('match', {
                    describe: 'regular expression. Only run tests that match',
                    string: true
                });
                break;
            case serveTaskName:
                yargsArgs.reset()
                    .option('port', {
                    description: 'the port to serve on should be the next argument (e.g. "--port 80")'
                })
                    .option('nobrowser', {
                    description: 'don\'t open a browser after initial bundle'
                });
                break;
            default:
                yargsArgs.reset(); // These commands have no special options.
                break;
        }
        yargsArgs.usage([
            'CONFIG FILES'
        ].concat(this.customConfigs.map(function (config) {
            return "   " + config.configFile + os.EOL + (config.readmeUrl ? config.readmeUrl : '') + os.EOL;
        }))
            .join(os.EOL));
        // Note this overrides the getters for ship and production on args
        exports.args.ship = exports.args.production = (exports.args.production || exports.args.ship);
        gulpCoreBuild.setConfig({
            production: exports.args.production
        });
        var tasks = this.getTasks();
        tasks.forEach(function (executable, name) {
            gulp_core_build_1.task(name, executable);
        });
        // If they are just looking for the task list, save time by not reading config files
        if (!exports.args.tasks) {
            // Configure tasks for a normal build
            console.log('Build target: ' + (exports.args.ship ? shipBuildMode.toUpperCase() : debugBuildMode.toUpperCase()));
            this.setupSharedConfig(exports.args);
            this._loadCustomConfig(exports.args);
            this.finalizeSharedConfig(exports.args);
        }
        gulpCoreBuild.initialize(gulp);
    };
    SpWebBuild.prototype.getYargs = function () {
        var yargsArgs = yargs.usage()
            .option('production', {
            alias: 'p',
            describe: 'build in ship mode with full localization and minimization',
            boolean: true
        })
            .option('ship', {
            alias: 'p',
            describe: 'build in ship mode with full localization and minimization',
            boolean: true
        })
            .option('locale', {
            alias: 'l',
            describe: 'override the default culture (e.g. "fr-fr")',
            string: true
        })
            .option('verbose', {
            describe: 'run the build with verbose logging'
        })
            .option('tasks', {
            alias: ['T', 'tasks-simple'],
            describe: 'shows the list of tasks which can be run'
        })
            .help('h')
            .global(['production', 'locale', 'verbose', 'h'])
            .command(buildTaskName, 'build the project')
            .command(bundleTaskName, 'build, localize, and bundle the project')
            .command(deployAzureStorageTaskName, 'upload the assets to a development CDN')
            .command(packageSolutionTaskName, 'package the project into a SPAPP')
            .command(testTaskName, 'build, localize, and bundle the project and run tests, and verify the coverage')
            .command(serveTaskName, 'build and bundle the project and run the development server')
            .command(defaultTaskName, 'equivalent to bundle');
        return yargsArgs;
    };
    /**
     * Define default task groups.
     */
    SpWebBuild.prototype.getTasks = function () {
        var result = new Map();
        result.set(buildTaskName, this.getBuildTask());
        result.set(bundleTaskName, this.getBundleTask());
        result.set(testTaskName, this.getTestTask());
        result.set(deployAzureStorageTaskName, spBuildCoreTasks.deployAzureStorage);
        result.set(packageSolutionTaskName, spBuildCoreTasks.packageSolution);
        // @todo VSO #167343 - remove this once bug from running 'gulp bundle test' in @microsoft/gulp-core-build is fixed
        result.set(testTaskName, this.getTestTask());
        result.set(serveTaskName, gulp_core_build_1.serial(result.get(bundleTaskName), gulp_core_build_serve_1.serve, gulp_core_build_1.watch(['src/**/*.{ts,tsx,scss,resx,js,json,html}',
            '!src/**/*.{scss.ts,resx.ts}'], gulp_core_build_1.serial(result.get(bundleTaskName), gulp_core_build_serve_1.reload))));
        result.set(defaultTaskName, result.get(bundleTaskName));
        return result;
    };
    SpWebBuild.prototype.getTestTask = function () {
        return gulp_core_build_1.serial(this.getBundleTask(), (enableCasperTests
            ? gulp_core_build_1.parallel(spBuildCoreTasks.casperJSRunner, gulp_core_build_karma_1.default)
            : gulp_core_build_karma_1.default));
    };
    SpWebBuild.prototype.getCoreBuildTask = function () {
        return gulp_core_build_1.parallel(gulp_core_build_typescript_1.tslint, gulp_core_build_1.serial(gulp_core_build_sass_1.default, gulp_core_build_1.serial(gulp_core_build_typescript_1.typescript, gulp_core_build_typescript_1.removeTripleSlash), gulp_core_build_1.serial.apply(void 0, additionalBuildTasks)), spBuildCoreTasks.copyStaticAssets);
    };
    SpWebBuild.prototype.getBuildTask = function () {
        return gulp_core_build_1.serial(exports.preCopy, this.getCoreBuildTask(), exports.postCopy);
    };
    SpWebBuild.prototype.getBundleTask = function () {
        return gulp_core_build_1.serial(this.getBuildTask(), spBuildCoreTasks.configureWebpack, gulp_core_build_webpack_1.default, spBuildCoreTasks.prepareDeploy, spBuildCoreTasks.writeManifests);
    };
    SpWebBuild.prototype.setupSharedConfig = function (buildArgs) {
        sp_tslint_rules_1.initializeTslintTask(gulp_core_build_typescript_1.tslint);
        gulp_core_build_sass_1.default.setConfig({
            dropCssFiles: true // Drops .css files in the lib directory for webpack
        });
        gulp_core_build_karma_1.default.setConfig({
            karmaConfigPath: path.resolve(path.join(__dirname, '..', 'karma.config.js'))
        });
        spBuildCoreTasks.configureWebpack.setConfig({
            webpack: gulp_core_build_webpack_1.default
        });
        spBuildCoreTasks.writeManifests.setConfig({
            debugLocale: buildArgs.locale
        });
        if (buildArgs.production) {
            spBuildCoreTasks.copyAssets.setConfig({
                extsToIgnore: ['.map', '.stats.json', '.stats.html']
            });
        }
    };
    /**
     * This function cleans up the shared config by populating task config properties that depend on other tasks'
     *  user-defined properties.
     */
    SpWebBuild.prototype.finalizeSharedConfig = function (buildArgs) {
        spBuildCoreTasks.writeManifests.setConfig({
            deployCdnPath: spBuildCoreTasks.copyAssets.taskConfig.deployCdnPath,
            debugBasePath: (gulp_core_build_serve_1.serve.taskConfig.https ? 'https' : 'http') + "://" + (os.hostname() || 'localhost') + ":" + gulp_core_build_serve_1.serve.taskConfig.port + "/"
        });
        if (buildArgs.production) {
            spBuildCoreTasks.packageSolution.setConfig({
                paths: {
                    manifestDir: spBuildCoreTasks.copyAssets.taskConfig.deployCdnPath
                }
            });
        }
    };
    SpWebBuild.prototype._loadCustomConfig = function (buildArgs) {
        // Read config files from project folders for various projects
        // @todo VSO #178074 - move this into each task, respectively
        // Load config file for each plugin
        for (var _i = 0, _a = this.customConfigs; _i < _a.length; _i++) {
            var taskConfigInfo = _a[_i];
            var configFilename = path.join(process.cwd(), 'config', taskConfigInfo.configFile);
            var schemaFilename = taskConfigInfo.schemaFile;
            var rawConfig = this._readConfigFile(buildArgs, configFilename, schemaFilename, taskConfigInfo.readmeUrl);
            if (rawConfig) {
                taskConfigInfo.callback(rawConfig);
            }
        }
    };
    SpWebBuild.prototype._readConfigFile = function (buildArgs, filename, schemafile, helpUrl) {
        if (!fs.existsSync(filename)) {
            return undefined;
        }
        else {
            if (buildArgs.verbose) {
                console.log("Found config file: " + path.basename(filename));
            }
            var contents = fs.readFileSync(filename);
            var stripped = stripJsonComments(contents.toString());
            var rawConfig = JSON.parse(stripped);
            delete rawConfig[schemaKey];
            if (schemafile) {
                var schema = require(schemafile);
                if (!this._schemaValidator.validate(rawConfig, schema)) {
                    var error = this._schemaValidator.getLastError();
                    var errorMessage = ("Error parsing file '" + path.basename(filename) + "', section [" + error.path + "]:")
                        + os.EOL + ("(" + error.code + ") " + error.message + " ") + os.EOL + os.EOL +
                        (helpUrl ? 'For more usage, please read:' + os.EOL + helpUrl : '');
                    console.log(os.EOL + 'ERROR: ' + errorMessage + os.EOL + os.EOL);
                    throw new Error(errorMessage);
                }
            }
            return rawConfig;
        }
    };
    return SpWebBuild;
}());
exports.SpWebBuild = SpWebBuild;
// Replace gcb.initialize to call our init stuff first.
exports.initialize = function (gulp) {
    var build = new SpWebBuild();
    build.initialize(gulp);
};
