// @microsoft/sp-tslint-rules/lib/rules/a11yRoleSupportsAriaPropsRule.js
/**
 * @copyright Microsoft Corporation. All rights reserved.
 *
 * @a11yRoleSupportsAriaPropsRule tslint rule of accessibility.
 */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var ts = require('typescript');
var Lint = require('tslint/lib/lint');
var getImplicitRole_1 = require('../utils/getImplicitRole');
var JsxAttribute_1 = require('../utils/JsxAttribute');
var roles = require('../utils/attributes/roleSchema.json');
var ariaAttributes = require('../utils/attributes/ariaSchema.json');
var roleString = 'role';
function getFailureStringForNotImplicitRole(roleNamesInElement, invalidPropNames) {
    return ("Attribute(s) " + invalidPropNames.join(', ') + " are not supported by role(s) " + roleNamesInElement.join(', ') + ". ") +
        "You are using incorrect role or incorrect aria-* attribute";
}
exports.getFailureStringForNotImplicitRole = getFailureStringForNotImplicitRole;
function getFailureStringForImplicitRole(tagName, roleName, invalidPropNames) {
    return ("Attribute(s) " + invalidPropNames.join(', ') + " not supported ") +
        ("by role " + roleName + " which is implicitly set by the HTML tag " + tagName + ".");
}
exports.getFailureStringForImplicitRole = getFailureStringForImplicitRole;
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (sourceFile) {
        return sourceFile.languageVariant === ts.LanguageVariant.JSX
            ? this.applyWithWalker(new A11yRoleSupportsAriaPropsWalker(sourceFile, this.getOptions()))
            : [];
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var A11yRoleSupportsAriaPropsWalker = (function (_super) {
    __extends(A11yRoleSupportsAriaPropsWalker, _super);
    function A11yRoleSupportsAriaPropsWalker() {
        _super.apply(this, arguments);
    }
    A11yRoleSupportsAriaPropsWalker.prototype.visitJsxElement = function (node) {
        this.checkJsxElement(node.openingElement);
        _super.prototype.visitJsxElement.call(this, node);
    };
    A11yRoleSupportsAriaPropsWalker.prototype.visitJsxSelfClosingElement = function (node) {
        this.checkJsxElement(node);
        _super.prototype.visitJsxSelfClosingElement.call(this, node);
    };
    A11yRoleSupportsAriaPropsWalker.prototype.checkJsxElement = function (node) {
        var attributesInElement = JsxAttribute_1.getJsxAttributesFromJsxElement(node);
        var roleProp = attributesInElement[roleString];
        // If role attribute is specified, get the role value. Otherwise get the implicit role from tag name.
        var roleValue = roleProp ? JsxAttribute_1.getStringLiteral(roleProp) : getImplicitRole_1.getImplicitRole(node);
        var isImplicitRole = !roleProp && !!roleValue;
        var normalizedRoles = (roleValue || '').toLowerCase().split(' ')
            .filter(function (role) { return !!roles[role]; });
        if (normalizedRoles.length === 0) {
            return;
        }
        var supportedAttributeNames = [];
        normalizedRoles.forEach(function (role) {
            supportedAttributeNames = supportedAttributeNames.concat(roles[role].supportedProps || []);
        });
        var attributeNamesInElement = Object.keys(attributesInElement)
            .filter(function (attributeName) { return !!ariaAttributes[attributeName.toLowerCase()]; });
        // Get the list of not-supported aria-* attributes in current element.
        var invalidAttributeNamesInElement = attributeNamesInElement
            .filter(function (attributeName) { return supportedAttributeNames.indexOf(attributeName) === -1; });
        if (invalidAttributeNamesInElement.length > 0) {
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), isImplicitRole ?
                getFailureStringForImplicitRole(node.tagName.getText(), normalizedRoles[0], invalidAttributeNamesInElement) :
                getFailureStringForNotImplicitRole(normalizedRoles, invalidAttributeNamesInElement)));
        }
    };
    return A11yRoleSupportsAriaPropsWalker;
}(Lint.RuleWalker));
