// @microsoft/sp-tslint-rules/lib/rules/a11yRoleHasRequiredAriaPropsRule.js
/**
 * @copyright Microsoft Corporation. All rights reserved.
 *
 * @a11yRoleHasRequiredAriaPropsRule tslint rule of accessibility.
 */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var ts = require('typescript');
var Lint = require('tslint/lib/lint');
var getImplicitRole_1 = require('../utils/getImplicitRole');
var JsxAttribute_1 = require('../utils/JsxAttribute');
var roles = require('../utils/attributes/roleSchema.json');
var ariaAttributes = require('../utils/attributes/ariaSchema.json');
var roleString = 'role';
function getFailureStringForNotImplicitRole(roleNamesInElement, missingProps) {
    return ("Element with ARIA role(s) '" + roleNamesInElement.join(', ') + "' ") +
        ("are missing required attribute(s): " + missingProps.join(', ') + ". ") +
        "A reference to role definitions can be found at https://www.w3.org/TR/wai-aria/roles#role_definitions.";
}
exports.getFailureStringForNotImplicitRole = getFailureStringForNotImplicitRole;
function getFailureStringForImplicitRole(tagName, roleNamesInElement, missingProps) {
    return ("Tag '" + tagName + "' has implicit role '" + roleNamesInElement + "'. ") +
        ("It requires aria-* attributes: " + missingProps.join(', ') + " that are missing in the element. ") +
        "A reference to role definitions can be found at https://www.w3.org/TR/wai-aria/roles#role_definitions.";
}
exports.getFailureStringForImplicitRole = getFailureStringForImplicitRole;
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (sourceFile) {
        return sourceFile.languageVariant === ts.LanguageVariant.JSX
            ? this.applyWithWalker(new A11yRoleHasRequiredAriaPropsWalker(sourceFile, this.getOptions()))
            : [];
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var A11yRoleHasRequiredAriaPropsWalker = (function (_super) {
    __extends(A11yRoleHasRequiredAriaPropsWalker, _super);
    function A11yRoleHasRequiredAriaPropsWalker() {
        _super.apply(this, arguments);
    }
    A11yRoleHasRequiredAriaPropsWalker.prototype.visitJsxElement = function (node) {
        this.checkJsxElement(node.openingElement);
        _super.prototype.visitJsxElement.call(this, node);
    };
    A11yRoleHasRequiredAriaPropsWalker.prototype.visitJsxSelfClosingElement = function (node) {
        this.checkJsxElement(node);
        _super.prototype.visitJsxSelfClosingElement.call(this, node);
    };
    A11yRoleHasRequiredAriaPropsWalker.prototype.checkJsxElement = function (node) {
        var attributesInElement = JsxAttribute_1.getJsxAttributesFromJsxElement(node);
        var roleProp = attributesInElement[roleString];
        // If role attribute is specified, get the role value. Otherwise get the implicit role from tag name.
        var roleValue = roleProp ? JsxAttribute_1.getStringLiteral(roleProp) : getImplicitRole_1.getImplicitRole(node);
        var isImplicitRole = !roleProp && !!roleValue;
        var normalizedRoles = (roleValue || '').toLowerCase().split(' ')
            .filter(function (role) { return !!roles[role]; });
        if (normalizedRoles.length === 0) {
            return;
        }
        var requiredAttributeNames = [];
        normalizedRoles.forEach(function (role) {
            requiredAttributeNames = requiredAttributeNames.concat(roles[role].requiredProps || []);
        });
        var attributeNamesInElement = Object.keys(attributesInElement)
            .filter(function (attributeName) { return !!ariaAttributes[attributeName.toLowerCase()]; });
        // Get the list of missing required aria-* attributes in current element.
        var missingAttributes = requiredAttributeNames
            .filter(function (attributeName) { return attributeNamesInElement.indexOf(attributeName) === -1; });
        if (missingAttributes.length > 0) {
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), isImplicitRole ?
                getFailureStringForImplicitRole(node.tagName.getText(), normalizedRoles[0], missingAttributes) :
                getFailureStringForNotImplicitRole(normalizedRoles, missingAttributes)));
        }
    };
    return A11yRoleHasRequiredAriaPropsWalker;
}(Lint.RuleWalker));
